libnmstate/nm/sriov.py

```diff
@@ -162,42 +148,66 @@ def get_info(device):
     except FileNotFoundError:
         return info
 
-    connection = nm_connection.ConnectionProfile()
-    connection.import_by_device(device)
-    if not connection.profile:
-        info[Ethernet.SRIOV_SUBTREE] = sriov_config
-        return info
+    if sriov_config[Ethernet.SRIOV.TOTAL_VFS]:
+        sriov_config[Ethernet.SRIOV.VFS_SUBTREE] = _get_sriov_vfs_info(ifname)
+    else:
+        sriov_config[Ethernet.SRIOV.VFS_SUBTREE] = []
 
-    sriov_setting = connection.profile.get_setting_by_name(
-        nmclient.NM.SETTING_SRIOV_SETTING_NAME
-    )
-
-    if sriov_setting:
-        vfs_config = _get_info_sriov_vfs_config(sriov_setting)
-        sriov_config[Ethernet.SRIOV.VFS_SUBTREE] = vfs_config
-
-        info[Ethernet.SRIOV_SUBTREE] = sriov_config
+    info[Ethernet.SRIOV_SUBTREE] = sriov_config
 
     return info
 
 
-def _get_info_sriov_vfs_config(sriov_setting):
-    vfs_config = []
-    vfs_setting = sriov_setting.props.vfs
-    for vf in vfs_setting:
-        vf_config = {}
-        vf_config[Ethernet.SRIOV.VFS.ID] = vf.get_index()
-        for nm_attribute in SRIOV_NM_TO_NMSTATE_MAP.keys():
-            _get_nm_attribute(vf, vf_config, nm_attribute)
+def _get_sriov_vfs_info(ifname):
+    """
+    This is a workaround to get the VFs configuration from runtime.
+    Ref: https://bugzilla.redhat.com/1777520
+    """
+    proc = subprocess.run(
+        ("ip", "link", "show", ifname),
+        stdout=subprocess.PIPE,
+        encoding="utf-8",
+    )
+    iplink_output = proc.stdout
 
-        vfs_config.append(vf_config)
+    # This is ignoring the first two line of the ip link output because they
+    # are about the PF and we don't need them.
+    vfs = iplink_output.splitlines(False)[2:]
+    vfs_config = [
+        vf_config for vf_config in _parse_ip_link_output_for_vfs(vfs)
+    ]
 
     return vfs_config
 
 
-def _get_nm_attribute(vf_object, vf_config, nm_attribute):
-    nmstate_key, nm_variant = SRIOV_NM_TO_NMSTATE_MAP[nm_attribute]
-    if vf_object.get_attribute(nm_attribute) is not None:
-        vf_config[nmstate_key] = nm_variant(
-            vf_object.get_attribute(nm_attribute)
-        )
+def _parse_ip_link_output_for_vfs(vfs):
+    vf_id = 0
+    for vf in vfs:
+        vf_config = {
+            Ethernet.SRIOV.VFS.ID: vf_id,
+        }
+        vf_config.update(_parse_ip_link_output_options_for_vf(vf))
+        vf_id += 1
+        yield vf_config
+
+
+def _parse_ip_link_output_options_for_vf(vf):
+    vf_options = {}
+    for option, expr in SRIOV_NMSTATE_TO_REGEX.items():
+        option_expr = re.compile(expr)
```

Comment:
You can store the compiled expression in the map and not recompile it each time.

Reply:
Done, thanks!